src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/HttpHandlerDiagnosticListener.cs

```diff
@@ -0,0 +1,370 @@
+                                {
+                                    newTable[entry.Key] = entry.Value;
+                                }
+                            }
+                            s_connectionGroupListField.SetValue(servicePoint, newTable);
+                        }
+                    }
+
+                    base[key] = value;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Helper class used for ServicePoint.m_ConnectiopGroupList. The goal here is to
+        /// intercept each new ConnectionGroup object being added to ServicePoint.m_ConnectionGroupList
+        /// and replace its m_ConnectionList arraylist field.
+        /// </summary>
+        private class ConnectionGroupHashtable : Hashtable
+        {
+            public override object this[object key]
+            {
+                get
+                {
+                    return base[key];
+                }
+                set
+                {
+                    if (s_connectionGroupType.IsInstanceOfType(value))
+                    {
+                        // Replace the Connection arraylist inside this ConnectionGroup object,
+                        // which allows us to intercept each new Connection object added under
+                        // this ConnectionGroup.
+                        ArrayList originalArrayList = s_connectionListField.GetValue(value) as ArrayList;
+                        ConnectionArrayList newArrayList = new ConnectionArrayList();
+                        if (originalArrayList != null)
+                        {
+                            foreach (object entry in originalArrayList)
+                            {
+                                newArrayList.Add(entry);
+                            }
+                        }
+                        s_connectionListField.SetValue(value, newArrayList);
+                    }
+
+                    base[key] = value;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Helper class used for ConnectionGroup.m_ConnectionList. The goal here is to
+        /// intercept each new Connection object being added to ConnectionGroup.m_ConnectionList
+        /// and replace its m_WriteList arraylist field.
+        /// </summary>
+        private class ConnectionArrayList : ArrayList
+        {
+            public override int Add(object value)
+            {
+                if (s_connectionType.IsInstanceOfType(value))
+                {
+                    // Replace the HttpWebRequest arraylist inside this Connection object,
+                    // which allows us to intercept each new HttpWebRequest object added under
+                    // this Connection.
+                    ArrayList originalArrayList = s_writeListField.GetValue(value) as ArrayList;
+                    HttpWebRequestArrayList newArrayList = new HttpWebRequestArrayList();
```

Comment:
Similar to my comment on initializing ```ConnectionArrayList```, add a constructor that does the ```AddRange``` for you.